server/src/main/java/org/opensearch/cluster/routing/IndexShardRoutingTable.java

```diff
@@ -292,6 +292,74 @@ public ShardIterator activeInitializingShardsRankedIt(
         return new PlainShardIterator(shardId, ordered);
     }
 
+    /**
+     * Returns an iterator over active and initializing shards, shards are ordered by weighted round-robin scheduling
+     * policy.
+     * @param wrrWeight Weighted round-robin weight entity
+     * @param nodes discovered nodes in the cluster
+     * @return an iterator over active and initializing shards, ordered by weighted round-robin
+     * scheduling policy. Making sure that initializing shards are the last to iterate through.
+     */
+    public ShardIterator activeInitializingShardsWRR(WRRWeights wrrWeight, DiscoveryNodes nodes, WRRShardsCache cache) {
+        final int seed = shuffler.nextSeed();
+        List<ShardRouting> ordered = new ArrayList<>(activeShards.size() + allInitializingShards.size());
+        List<ShardRouting> orderedActiveShards;
+        if (cache.get(new WRRShardsCache.Key(shardId)) != null) {
+            orderedActiveShards = cache.get(new WRRShardsCache.Key(shardId));
+        } else {
+            orderedActiveShards = getShardsWRR(activeShards, wrrWeight, nodes);
+            cache.put(new WRRShardsCache.Key(shardId), orderedActiveShards);
+        }
+        ordered.addAll(shuffler.shuffle(orderedActiveShards, seed));
+
+        if (!allInitializingShards.isEmpty()) {
+            List<ShardRouting> orderedInitializingShards = getShardsWRR(allInitializingShards, wrrWeight, nodes);
+            ordered.addAll(orderedInitializingShards);
+        }
+        return new PlainShardIterator(shardId, ordered);
+    }
+
+    /**
+     *
+     * @param shards shards to be ordered using weighted round-robin scheduling policy
+     * @param wrrWeight weights to be considered for routing
+     * @param nodes discovered nodes in the cluster
+     * @return list of shards ordered using weighted round-robin scheduling.
+     */
+    private List<ShardRouting> getShardsWRR(List<ShardRouting> shards, WRRWeights wrrWeight, DiscoveryNodes nodes) {
+        List<WeightedRoundRobin.Entity<ShardRouting>> weightedShards = calculateShardWeight(shards, wrrWeight, nodes);
+        WeightedRoundRobin<ShardRouting> wrr = new WeightedRoundRobin<>(weightedShards);
+        List<WeightedRoundRobin.Entity<ShardRouting>> wrrOrderedActiveShards = wrr.orderEntities();
+        List<ShardRouting> orderedActiveShards = new ArrayList<>(activeShards.size());
+        for (WeightedRoundRobin.Entity<ShardRouting> shardRouting : wrrOrderedActiveShards) {
+            orderedActiveShards.add(shardRouting.getTarget());
+        }
+        return orderedActiveShards;
+    }
+
+    /**
+     *
+     * @param shards associate weights to shards
+     * @param wrrWeight weights to be used for association
```

Comment:
nit add a description, make this package private and please add tests

Reply:
Sure, will add a description. The method is already private. Do you mean we should move it? Tests added in OperationRoutingTests.java do call these methods. Do you think we should have separate tests as well?

Reply:
yes these are critical operations and we might need more tests

Reply:
added tests.